[Clang][Sema] Print more static_assert exprs

sethp opened the pull request:
This change introspects more values involved in a static_assert, and extends the supported set of operators for introspection to include binary operator method calls.

It's intended to address the use-case where a small static_assert helper looks something like this (via `constexpr-builtin-bit-cast.cpp`):

```c++
struct int_splicer {
  unsigned x;
  unsigned y;

  constexpr bool operator==(const int_splicer &other) const {
    return other.x == x && other.y == y;
  }
};
```

When used like so:

```c++
constexpr int_splicer got{1, 2};
constexpr int_splicer want{3, 4};
static_assert(got == want);
```

Then we'd expect to get the error:

```
Static assertion failed due to requirement 'got == want'
```

And this change adds the helpful note:

```
Expression evaluates to '{1, 2} == {3, 4}'
```

sethp committed:
[Clang][Sema] Print more static_assert exprs

This change introspects more values involved in a static_assert, and
extends the supported set of operators for introspection to include
binary operator method calls.

It's intended to address the use-case where a small static_assert helper
looks something like this (via `constexpr-builtin-bit-cast.cpp`):

```c++
struct int_splicer {
  unsigned x;
  unsigned y;

  constexpr bool operator==(const int_splicer &other) const {
    return other.x == x && other.y == y;
  }
};
```

When used like so:

```c++
constexpr int_splicer got{1, 2};
constexpr int_splicer want{3, 4};
static_assert(got == want);
```

Then we'd expect to get the error:

```
Static assertion failed due to requirement 'got == want'
```

And this change adds the helpful note:

```
Expression evaluates to '{1, 2} == {3, 4}'
```

shafik reviewed (commented):
I don't feel like the changes are an improvement to the diagnostic but I would like to hear from others who focus on clang front-end reviews.

tahonermann commented:
This looks like a good change to me since coercing a compiler to reveal the values of compile-time objects can be frustrating. We could really use something like `constexpr_print` from [P2758 (Emitting messages at compile time)](https://wg21.link/p2758).

tbaederr commented:
I like this change, the only thing I'm a little afraid of is that the output will deteriorate for large structs

sethp commented:
Thanks for the feedback! I've definitely found it helpful trying to puzzle out why one of my static_asserts is failing, though it's certainly a far less complete solution than the document @tahonermann shared. 

The output from `printPretty` will definitely fall apart for a complex structure (especially one with e.g. multiple bases, or lots of similar fields). Still, this change seems like an overall improvement to me: it's still quite possible to cut a complex struct comparison up into multiple `static_assert`s. Right now I feel more or less compelled to cut every comparison down to asserts on built-in types, and this output was intended to raise that floor up to "a struct with a handful of fields and an `operator==`." 

That said, I'm receiving approximately $0 from Big `printPretty` lobbying groups, so I'm happy to walk the structure however we feel `ConvertAPValueToString` ought to do the conversion. I do quite like how `clangd` does its inline initializer hints, so maybe something like:

```
Expression evaluates to '{x: 1, y: 2} == {x: 3, y: 4}'
```

for the original example, or for a `int[2]`:

```
Expression evaluates to '{[0]=1, [1]=2} == ...'
```

What do you think?

erichkeane reviewed (commented):
I too would like to see some more complicated dumps for 'complicated' structs (inheritance, multiple inheritance, etc), and vectors/arrays of complicated structs.

I am really concerned with the 'depth' that this will show and be unwieldy as a result.

sethp commented on clang/test/CXX/class/class.compare/class.eq/p3.cpp:
```diff
@@ -6,20 +6,20 @@ struct A {
-static_assert(A{1, 2, 3, 4, 5} == A{1, 2, 0, 4, 5}); // expected-error {{failed}}
-static_assert(A{1, 2, 3, 4, 5} == A{1, 2, 3, 0, 5}); // expected-error {{failed}}
-static_assert(A{1, 2, 3, 4, 5} == A{1, 2, 3, 4, 0}); // expected-error {{failed}}
+static_assert(A{1, 2, 3, 4, 5} == A{0, 2, 3, 4, 5}); // expected-error {{failed}} expected-note {{evaluates to}}
```
I can, though I was worried it'd over-specify advisory output to do so. What do you think about waiting until we reached loose consensus around the format for printing the struct? 

If you're looking for a sample here's a snippet of output from my `build/bin/clang clang/test/CXX/class/class.compare/class.eq/p3.cpp` :

```
clang/test/CXX/class/class.compare/class.eq/p3.cpp:9:15: error: static assertion failed due to requirement 'A{1, 2, 3, 4, 5} == A{0, 2, 3, 4, 5}'
    9 | static_assert(A{1, 2, 3, 4, 5} == A{0, 2, 3, 4, 5}); // expected-error {{failed}} expected-note {{evaluates to}}
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
clang/test/CXX/class/class.compare/class.eq/p3.cpp:9:32: note: expression evaluates to '{1, {2, 3, 4}, 5} == {0, {2, 3, 4}, 5}'
    9 | static_assert(A{1, 2, 3, 4, 5} == A{0, 2, 3, 4, 5}); // expected-error {{failed}} expected-note {{evaluates to}}
      |               ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
clang/test/CXX/class/class.compare/class.eq/p3.cpp:10:15: error: static assertion failed due to requirement 'A{1, 2, 3, 4, 5} == A{1, 0, 3, 4, 5}'
   10 | static_assert(A{1, 2, 3, 4, 5} == A{1, 0, 3, 4, 5}); // expected-error {{failed}} expected-note {{evaluates to}}
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
clang/test/CXX/class/class.compare/class.eq/p3.cpp:10:32: note: expression evaluates to '{1, {2, 3, 4}, 5} == {1, {0, 3, 4}, 5}'
   10 | static_assert(A{1, 2, 3, 4, 5} == A{1, 0, 3, 4, 5}); // expected-error {{failed}} expected-note {{evaluates to}}
```

That's for:

```c++
struct A {
  int a, b[3], c;
  bool operator==(const A&) const = default;
};
```

Which, to my eye, is a strict improvement over the only alternative (without this patch) that I see to get diagnostics:

```c++
// instead of static_assert(A{1, 2, 3, 4, 5} == A{0, 2, 3, 4, 5});
constexpr A lhs = {1, 2, 3, 4, 5};
constexpr A rhs = {0, 2, 3, 4, 5};
static_assert(lhs.a == rhs.a);
static_assert(lhs.b[0] == rhs.b[0]);
static_assert(lhs.b[1] == rhs.b[1]);
static_assert(lhs.b[2] == rhs.b[2]);
static_assert(lhs.c == rhs.c);
```

sethp replied:
Actually, on second thought, I think I'd prefer to leave these under-specified (since they're really checking the behavior of the default `operator==`).

I did go ahead and add a test specifically for some more complicated types in [2fd5fc4](https://github.com/llvm/llvm-project/pull/74852/commits/2fd5fc464868a682cb4b66ea4e97a6a7100ab120) , though, to give us a few specific examples to focus on for discussion

erichkeane replied:
The additional test is sufficient to me I think.

sethp committed:
add test for review discussion

sethp committed:
fixup!: remove notes to self unrelated to PR

Oops, those were just things I was curious about, not meant to be left
in.

erichkeane commented on clang/test/SemaCXX/static-assert-diagnostics.cpp:
```diff
@@ -0,0 +1,117 @@
+// maybe it's ok? I mean, not like we can do better really...
+constexpr auto _ = (int[3]){2, 3, 4};
+
+// output: '{{2, 3, 4}} == {0, 3, 4}'  (the `{{` breaks VerifyDiagnosticConsumer::ParseDirective)
```
IIRC, you can escape a { with \.

sethp replied:
It didn't seem to like it; when I tried 

```
// expected-note@+1 {{evaluates to '\{{2, 3, 4}\} == {0, 3, 4}'}}
```

here I got this output from `-verify`:

```
error: 'expected-error' diagnostics seen but not expected: 
  File /home/seth/Code/src/github.com/llvm/llvm-project/clang/test/SemaCXX/static-assert-diagnostics.cpp Line 30: cannot find end ('}}') of expected string
error: 'expected-note' diagnostics seen but not expected: 
  File /home/seth/Code/src/github.com/llvm/llvm-project/clang/test/SemaCXX/static-assert-diagnostics.cpp Line 31: expression evaluates to '{{2, 3, 4}} == {0, 3, 4}'
2 errors generated.
```

I *think* I could use the `-re` mode, but then `{{` has _meaning_ so the output would be really obscured, something like `{{[{][{]}}` would be how to match `{{`, I think.

My bad idea is to teach the parser to count opening braces instead, so a `expected-error {{{` looks for a `}}}` and ignores any `{{` or `}}` sequences, but I wanted to get this up before dealing with that particular sub-problem.

sethp replied:
Well, my bad idea [worked out easily enough](https://github.com/sethp/llvm-project/commit/153d36338a73d8fbbb6087b2cf07e671b7aa660a). I'm happy to roll that in to (or out of) this change: whatever y'all think is best.

(I also seem to have confused the 'hub with an immaculately timed interrupt to the `git push`; that commit is on the appropriate branch in my fork, but hasn't shown up here on the PR yet. Maybe it will eventually? ¯\\_(ツ)_/¯ )

erichkeane replied:
Please put that 'wide delimiters' thing as a separate pull request that we can review, I suspect it'll go quickly, but i want a few folks to chime in who have experience on the consumer.

sethp committed:
[Clang] Wide delimiters ('{{{') for expect strings

Prior to this commit, it was impossible to use the simple string
matching directives to look for most content that contains `{{`, such
as:

```
// expected-note {{my_struct{{1}, 2}}}
```

Which would parse like so:

```
             "nested" brace v
// expected-note {{my_struct{{1}, 2}}}
          closes the nested brace  ^ |
                            trailing }
```

And the frontend would complain 'cannot find end ('}}') of expected'.

At this snapshot, VerifyDiagnosticConsumer's parser now counts the
opening braces and looks for a matching length of closing sigils,
allowing the above to be written as:

```
// expected-note {{{my_struct{{1}, 2}}}}
   opening brace |-|                 |-|
  closing brace is '}}}', found here ^
```

sethp committed:
Revert "[Clang] Wide delimiters ('{{{') for expect strings"

This reverts commit 153d36338a73d8fbbb6087b2cf07e671b7aa660a.

sethp committed:
fixup!: Reapply "[Clang] Wide delimiters ('{{{') for expect strings"

This reverts commit 6597d6c3dda1890c0c999ef4fbc24e492a89ae8b.

sethp committed:
print a valid initializer expression for value

sethp commented:
In playing around with this some more, I found it was pretty effective for me to have an intializer-like expression (as in [e769f15](https://github.com/llvm/llvm-project/pull/74852/commits/e769f158c8ee001e95299ce4c83e5250e2a8f963) ), because then I could copy/paste the output from clang directly into a `static_assert()` in the source file and start cutting it down interactively with clangd's guidance and my editor identifying common subexpressions:

```c++
// evaluates to 'A{0, {0, 3, 4}, 5} == A{1, {2, 3, 4}, 5}' ->
static_assert(A{0, {0, 3, 4}, 5} == A{1, {2, 3, 4}, 5}); // ->
static_assert(A{0, {0}, 5} == A{1, {2}, 5}); // ->
static_assert(A{0, {0}} == A{1, {2}});
```

gets me to a minimal "diff" of the complex struct plus some nice paths to the different elements:

![image](https://github.com/llvm/llvm-project/assets/241129/3abe2c1b-1a19-4b3a-9fc6-32029c25ab40)

I could reproduce that in the textual output, but I worry that'd be too noisy without the dimming/color text effects.

sethp commented on clang/test/SemaCXX/static-assert-diagnostics.cpp:
```diff
@@ -0,0 +1,115 @@
+
+constexpr auto cc = C{A{1, {2, 3, 4}, 5}, B{7, 6}, C::E1};
+
+// expected-note@+1 {{{evaluates to 'C{{1, {2, 3, 4}, 5}, {7, 6}, 0} == C{{0, {0, 3, 4}, 5}, {5, 0}, 1}'}}}
+static_assert(cc == C{a0, {5}, C::E2}); // expected-error {{failed}}
```
NB: 

```c++
static_assert(C{{1, {2, 3, 4}, 5}, {7, 6}, 0} == C{{0, {0, 3, 4}, 5}, {5, 0}, 1});
//                                         ^                                  ^
```

doesn't work as an example of two fully initialized `C`s because the unnamed enum can't be initialized from the `int`s indicated above; if we like this idea, I'll look into how to print those out as `C::E1` and `C::E2`.

erichkeane replied:
I really like the idea of printing enum values (though we obviously have to revert to the number in cases where there isn't a value).

However, these longer ones make me wonder if we should be printing JUST the differences here instead of all of this.  I realize this is a divergence in the direction of this patch, but feels like it would be a MUCH nicer experience.

MAYBE it is a second note?  And perhaps it should only happen when there is no user-defined == (and maybe the printing everythign should only happen in that case?).

Consider:

```
struct S {
int a,b,c;
};
constexpr bool operator==(const S&, const S&) { return false;}

static_assert(S{1,2,3} == S{1,2,3});
```

This diagnostic would be REALLY confusing as is, right?  'static-assert-failed' followed by 'S{1,2,3} != S{1,2,3}`.

sethp replied:
It is confusing, but I don't think this change makes it any worse; the output I see from clang with my change is:

```c++
clang/test/SemaCXX/static-assert-diagnostics.cpp:177:15: error: static assertion failed due to requirement 'S{1, 2, 3} == S{1, 2, 3}'
  177 | static_assert(S{1,2,3} == S{1,2,3});
      |               ^~~~~~~~~~~~~~~~~~~~
clang/test/SemaCXX/static-assert-diagnostics.cpp:177:24: note: expression evaluates to 'S{1, 2, 3} == S{1, 2, 3}'
  177 | static_assert(S{1,2,3} == S{1,2,3});
      |               ~~~~~~~~~^~~~~~~~~~~
```

The `note` this change adds is redundant, sure, but I don't think it's more confusing than omitting it: either way, I'm going to go hunt for an `operator==` between two `S`s and see if I can figure out why two things that sure feel like they _ought_ to be `==` aren't.

A different example with the same `operator==` overload there might be something like this:

```
struct alignas(int) S {
unsigned char a, b, c;
};
constexpr bool operator==(const S&, const S&) { return false; }

static_assert(S{1, 2, 3} == std::bit_cast<S>(0x030201));
```

Which now outputs:

```
clang/test/SemaCXX/static-assert-diagnostics.cpp:177:15: error: static assertion failed due to requirement 'S{1, 2, 3} == std::bit_cast(197121)'
  177 | static_assert(S{1, 2, 3} == std::bit_cast<S>(0x030201));
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
clang/test/SemaCXX/static-assert-diagnostics.cpp:177:26: note: expression evaluates to 'S{1, 2, 3} == struct S{1, 2, 3}'
  177 | static_assert(S{1, 2, 3} == std::bit_cast<S>(0x030201));
      |               ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Here the note is, I think, quite helpful: to me it very strongly suggests to me that the issue isn't the `bit_cast` or a endian-ness mismatch or anything like that, but the `operator==` implementation.

sethp replied:
Oh, I also wanted to add: I like the idea of adding extra `note`s here when possible to be more helpful in diagnosing the problem. To me, that's what this PR does: adds more context that will help in at least a few situations that I've found.

erichkeane replied:
IMO, adding notes when it doesn't add additional information is a net-loss.  Folks already complain about the number of our error/notes when trying to figure something out, so I'm concerned this is actually making it worse when we don't print something 'different' in the note.

sethp replied:
I hear your frustration there: it's a very goldilocks problem to get the feedback "just right," and not one made easier by the, ah, breadth of individual preferences (and the cartesian product thereof with the standards).

In this case, would it make sense to mirror what clang does with e.g. integer constants, and skip the note if they evaluate identically to as-written? So, we'd avoid noting the redundant info for `static_assert(S{1, 2, 3} == S{1, 2, 3})`, but still emit the result of the `bit_cast` for `static_assert(S{1, 2, 3} == std::bit_cast<S>(0x030201))`?

erichkeane replied:
I think that would be a sensible improvement here. I think mixed with making sure we don't print arbitrarily sized structs/arrays would make this 'net improvement'.

sethp commented on clang/test/SemaCXX/static-assert-diagnostics.cpp:
```diff
@@ -0,0 +1,115 @@
+  }
+};
+
+// expected-note@+1 {{{evaluates to '_arr{{2, 3, 4}} == (int[3]){0, 3, 4}'}}}
+static_assert(_arr{2, 3, 4} == a0.b); // expected-error {{failed}}
```
this also doesn't quite work as advertised, either. The closest I could come was something like:

```
constexpr int v[3] = {0, 3, 4};
static_assert(_arr{{2, 3, 4}} == v);
// ->
constexpr int v[3] = {0};
static_assert(_arr{{2}} == v);
```

Because anything else and the evaluator complained about invalid `constexpr` subexpressions in:

```
In call to '_arr{{2}}.operator==(&(int[3]){0}[0])'
```

sethp replied:
Ah, happily this is resolved by doing less work: the `const` qualifier from `(const int [3]){0}` is semantically important here, and the type printer already "knows" how to print qualified types appropriately.

sethp committed:
[Clang][Sema] Print enumerators by name

Teaches APValue::printPretty to print appropriate cast expressions for
enumeral types rather than just bare ints, and to respect the
pre-existing PrintingPolicy::UseEnumerators field (falling back to
printing a cast when the value doesn't match any named constant).

sethp commented on clang/lib/AST/APValue.cpp:
```diff
@@ -711,11 +712,54 @@ void APValue::printPretty(raw_ostream &Out, const PrintingPolicy &Policy,
+      if (Policy.UseEnumerators) {
+        for (const EnumConstantDecl *ECD : ET->getDecl()->enumerators()) {
+          if (APSInt::isSameValue(ECD->getInitVal(), Val)) {
+            if (ECD->isCXXClassMember())
+              ECD->printQualifiedName(Out, Policy);
+            else
+              ECD->printName(Out, Policy);
```
This is a weird heuristic: without it, we were trying to print a function-local enum constant as `fn_name()::Constant`.

It might be better to go back to just always using the QualifiedName though, I'm not sure.

sethp committed:
fixup! [Clang][Sema] Print enumerators by name

sethp commented on clang/test/SemaCXX/static-assert-diagnostics.cpp:
```diff
@@ -0,0 +1,205 @@
+} // namespace vector
+
+namespace {
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+constexpr auto bits = 0x030201;
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+constexpr auto bits = 0x01020300;
+#else
+#error "don't know what to do with mixed endianness"
+#endif
+
+struct alignas(decltype(bits)) S {
+unsigned char a, b, c;
+};
+// confusing `==` on purpose
+constexpr bool operator==(const S&, const S&) { return false; }
+
+// the note should clearly implicate the `==` implementation
+// expected-note@+1 {{{evaluates to 'S{1, 2, 3} == S{1, 2, 3}'}}}
+static_assert(S{1, 2, 3} == std::bit_cast<S>(bits)); // expected-error {{failed}}
```
I went ahead and added this example to the test. 

Also, writing it served as a small experience report: I got the big-endian bits wrong at first. The little endian constant's padding is in the implicit leading zeros, so when I flipped it I wrote `0x010203`, and I saw:

```
clang/test/SemaCXX/static-assert-diagnostics.cpp:207:15: error: static assertion failed due to requirement '(anonymous namespace)::S{1, 2, 3} == std::bit_cast(bits)'
  207 | static_assert(S{1, 2, 3} == std::bit_cast<S>(bits)); // expected-error {{failed}}
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
clang/test/SemaCXX/static-assert-diagnostics.cpp:207:26: note: expression evaluates to 'S{1, 2, 3} == S{0, 1, 2}'
  207 | static_assert(S{1, 2, 3} == std::bit_cast<S>(bits)); // expected-error {{failed}}
      |               ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
```

the new note's output `'S{1, 2, 3} == S{0, 1, 2}'` effectively reminded me to add the padding bits to push the value "left."

There's plenty of cases where the new diagnostic breaks down, but it has been very helpful to me so far, and I wanted to make sure to highlight that.

sethp commented:
I think this change is useful enough on its own that I'd like us to consider merging it: to me, teaching the compiler to do additional diffing or further clarifying the structure of complicated sub-objects feel like they'd benefit from having this work as a fallback. We could then build up from concrete examples of "it would be nice if the compiler diff'd here", and rely on the much more capable (and domain-aware) programmers to use the complete evaluation result for solving the general case.

That said, I'm curious if there's an upper limit we ought to set here on the size of the string: copy & paste work fine for hundreds or even thousands of characters, but if someone `static_assert`s on a multi-GB `const char*`, they might be in for a bit of trouble.

erichkeane commented:
> I think this change is useful enough on its own that I'd like us to consider merging it: to me, teaching the compiler to do additional diffing or further clarifying the structure of complicated sub-objects feel like they'd benefit from having this work as a fallback. We could then build up from concrete examples of "it would be nice if the compiler diff'd here", and rely on the much more capable (and domain-aware) programmers to use the complete evaluation result for solving the general case.
> 

As I said before, I'm somewhat concerned this makes some situations worse by adding more noise, and is 'useless' in cases where the size is long enough to make the 'difference' invisible.  There is a judgement of value here that I'd probably like to hear the opinion of @AaronBallman and @cor3ntin .

> That said, I'm curious if there's an upper limit we ought to set here on the size of the string: copy & paste work fine for hundreds or even thousands of characters, but if someone `static_assert`s on a multi-GB `const char*`, they might be in for a bit of trouble.

We absolutely should limit the size of the notes, another concern I had though I thought we auto-snipped when printing large structures.  If that doesn't happen, we need to NOT do this patch, crashing a user's terminal is worse than giving them insufficient information.

sethp commented:
Yeah, definitely agree about adding no new issues like https://github.com/llvm/llvm-project/issues/71675 . The good news is that ought to make for an easy test case: something like `struct { _BitInt(__BITINT_MAXWIDTH__ >> 6) F }` should do it. I'll take a look!

sethp committed:
[Clang][Sema] Print limited, non-redundant vals

Introduces size_limited_ostream which clips output sent to it to a
maximum number of leading/trailing bytes.

Makes use of that new facility to avoid emitting overly large struct or
vector/array values (unless explicitly requested with the new
`-fconstexpr-print-value-size-limit`.

Additionally, look a little deeper in the expr tree to try and  avoid
printing redundant notes when the expressions evaluate exactly
as-written.

sethp committed:
fixup! [Clang][Sema] Print limited, non-redundant vals

erichkeane reviewed (commented):
I think this needs a 'merge' to get rid of the }}} patch changes from its diff, but I'm in favor of the patch otherwise.  Probably could use a release note, and a look over @cor3ntin or @cjdb  perhaps?

sethp committed:
Merge remote-tracking branch 'origin/main' into feat/static-assert-more-printing

Pick up baa8c2a to replace c270701 .

sethp committed:
fixup!: [doc] add release note

sethp committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into feat/static-assert-more-printing

sethp commented:
Looks like I missed the branch for v18 with this one, hence the conflict (now resolved), too bad.

The test that's failing is looking for clang to emit some options that it's not; it looks like the same test is failing on main, so I haven't dug too deep into it. I believe I'm just waiting on one more reviewer's approval (possibly @cor3ntin or @cjdb) for this change, and then I'll rebase/squash the history & it will be ready.

Does that sound right to y'all?

sethp committed:
Merge branch 'main' into feat/static-assert-more-printing

tbaederr commented:
Ping @cor3ntin @cjdb

tbaederr commented:
Ping @cor3ntin @cjdb

sethp committed:
Merge branch 'main' into feat/static-assert-more-printing

sethp commented:
Ping @cor3ntin @cjdb —this one definitely needs a rebase from me, but I don't want to do that if y'all have a review in progress.

sethp committed:
Update ReleaseNotes.rst

yronglin commented:
FYI, Does this PR have some overlapping functions with `__builtin_dump_struct` in structure printing? https://github.com/llvm/llvm-project/blob/db4170a4f3a701a62f5c1ef2e6a30f490f107f7d/clang/lib/Sema/SemaChecking.cpp#L453-L695

sethp commented:
@yronglin Thanks for introducing me to `__builtin_dump_struct`, that looks useful!

I'm not sure I see much in the way of overlap with this PR, though: would you say more about what you mean? Is there maybe some commonality that you're concerned might drift apart?

yronglin commented:
> @yronglin Thanks for introducing me to `__builtin_dump_struct`, that looks useful!
> 
> I'm not sure I see much in the way of overlap with this PR, though: would you say more about what you mean? Is there maybe some commonality that you're concerned might drift apart?

Sorry for not describing it clearly before, I really like this improvements. `__builtin_dump_struct` has the dilemma for complicated dumps for 'complicated' structs too. Can they have similar restricts and output formats in the future (not in this PR).

sethp commented:
Ah, I think I understand you: it sounds like you're imagining using `__builtin_dump_struct` to inspect/trace a value that ends up in a failing `static_assert`. Is that right?

I can definitely look into what combination of arguments to `__builtin_dump_struct` (if any) would produce similar output, that makes a lot of sense to me.

sethp committed:
Merge branch 'main' into feat/static-assert-more-printing

sethp commented:
Ping @cor3ntin @cjdb

sethp commented:
Ping @cor3ntin @cjdb

tbaederr commented:
Ping

cor3ntin commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -17172,6 +17179,113 @@ static void WriteCharValueForDiagnostic(uint32_t Value, const BuiltinType *BTy,
   }
 }
 
+/// A raw_ostream that clips its output beyond a certain size, inserting a
```
I this we should move that to llvm/Support - either in raw_ostream.h or a new file

cor3ntin commented on clang/include/clang/Basic/DiagnosticOptions.h:
```diff
@@ -84,6 +84,7 @@ class DiagnosticOptions : public RefCountedBase<DiagnosticOptions>{
     DefaultMacroBacktraceLimit = 6,
     DefaultTemplateBacktraceLimit = 10,
     DefaultConstexprBacktraceLimit = 10,
+    DefaultConstexprValueSizeLimit = 320,
```
320?

cor3ntin commented on clang/lib/AST/APValue.cpp:
```diff
@@ -711,11 +712,54 @@ void APValue::printPretty(raw_ostream &Out, const PrintingPolicy &Policy,
+                       Defn->getTypedefNameForAnonDecl()) {
+            assert(Typedef->getIdentifier() && "Typedef without identifier?");
+            Out << '(' << Typedef->getIdentifier()->getName() << ')';
+          } else if (const NamedDecl *ND = dyn_cast_if_present<NamedDecl>(
+                         Defn->getNextDeclInContext())) {
+            // if it's part of a declaration,  then use `(decltype(...))7`
+            Out << "(decltype(";
+            ND->printQualifiedName(Out);
+            Out << "))";
```
Does it really make sense to support the unnamed case?

cor3ntin commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -17256,6 +17378,56 @@ static bool ConvertAPValueToString(const APValue &V, QualType T,
+  case APValue::ValueKind::Struct: {
+    llvm::raw_svector_ostream OS(Str);
+    const auto *RT = T->getAsStructureType();
+    if (RT->hasUnnamedOrLocalType()) {
+      OS << '(';
+      // e.g. `(unnamed struct at ...)`, unless...
+      if (const auto *Defn = RT->getDecl()->getDefinition()) {
+        if (const NamedDecl *ND =
+                dyn_cast_if_present<NamedDecl>(Defn->getNextDeclInContext())) {
+          // ... it's part of a declaration,  then use `(decltype(...))`
+          OS << "decltype(";
+          ND->printQualifiedName(OS);
+          OS << ")";
+        } else
+          Defn->printQualifiedName(OS);
+      }
+      OS << ')';
```
There too, I think only supporting the named case would be a good start

cor3ntin commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -17256,6 +17378,56 @@ static bool ConvertAPValueToString(const APValue &V, QualType T,
+          Defn->printQualifiedName(OS);
+      }
+      OS << ')';
+    } else if (T.hasQualifiers()) {
+      OS << '(' << T << ')';
+    } else {
+      PrintingPolicy TyPolicy(Context.getPrintingPolicy());
+      TyPolicy.SuppressUnwrittenScope = true;
+      T.print(OS, TyPolicy);
+    }
```
Why are these two cases so different?

cor3ntin commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -17256,6 +17378,56 @@ static bool ConvertAPValueToString(const APValue &V, QualType T,
+    }
+    PrintingPolicy Policy(Context.getPrintingPolicy());
+    Policy.UseEnumerators = true;
+    unsigned Limit = Context.getDiagnostics().getConstexprValueSizeLimit();
+    if (Limit) {
+      size_limited_ostream OSS(OS, Limit);
+      V.printPretty(OSS, Policy, T, &Context);
+    } else
+      V.printPretty(OS, Policy, T, &Context);
```
You do that in a few places, can you create a function or lambda?
